[JENKINS-67237] BuildTrigger waits until the dependency graph has been updated - 2nd attempt

Si-So opened the pull request:
<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://www.jenkins.io/download/lts/#backporting-process for more.
-->
See [JENKINS-67237](https://issues.jenkins-ci.org/browse/JENKINS-67237).

The first [PR](https://github.com/jenkinsci/jenkins/pull/6101) caused this [regression](https://issues.jenkins.io/browse/JENKINS-68189). This PR is the 2nd attempt at solving [JENKINS-67237](https://issues.jenkins-ci.org/browse/JENKINS-67237) without causing a regression. 

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* When triggering a new build while the build graph is currently being re-computed, jenkins waits for the re-computation to finish. This gurantees that recently updated build triggers are executed.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/
-->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@[res0nance](https://github.com/res0nance)
@[timja](https://github.com/timja)

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Si-So committed:
Revert "Revert "[JENKINS-67237] `BuildTrigger` waits until the dependency graph has been updated (#6101)""

This reverts commit f3a32be3a3795664f4cdb01f92eaabe8a2f7ad3c.

Si-So committed:
[JENKINS-67237]: Test case that checks if ReverseBuildTriggers are fired after a restart of the jenkins instance

Si-So commented:
I created a TC to detect the [regression](https://issues.jenkins.io/browse/JENKINS-68189). I expect that only the new TC will fail during the currently running build. 

I have also prepared a fix for the regression that I will push once the currently running build is done.

Si-So committed:
[JENKINS-67237]: Fixes the regression from [JENKINS-68189]

Si-So commented:
As predicted only the newly added TC failed.

I have now pushed a fix for the regression and expect no failing unit tests during this run.

res0nance commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4920,15 +4956,37 @@ public void rebuildDependencyGraph() {
+
+            synchronized (dependencyGraphLock) {
+                //Wait for the currently running calculation to finish
+                if (calculatingFutureDependencyGraph != null) {
```
Because of how this works, potentially a call to `rebuildDependencyGraphAsync` will block which defeats the original purpose of the function. This needs a rethink

Si-So replied:
Thanks yet again for your feedback. I just pushed a new commit. 

I also noticed another bug: getFutureDependencyGraph does not return the most recent dependency graph if the most recent dependency graph has been calculated synchronously. This is problematic because BuildTrigger.execute assumes that getFutureDependencyGraph will always return the most recent dependency graph. To fix this problem calculatingFutureDependencyGraph is now set to null in scheduleCalculationOfFutureDependencyGraph after the calculation has finished.

Si-So committed:
[JENKINS-67237]: Small corrections in scheduleCalculationOfFutureDependencyGraph

1) scheduleCalculationOfFutureDependencyGraph does no longer potentially block rebuildDependencyGraphAsync
2) getFutureDependencyGraph returns the most recent dependency graph calculated by either rebuildDependencyGraphAsync or rebuildDependencyGraph

res0nance reviewed (approved):
LGTM

basil committed:
Merge remote-tracking branch 'origin/master' into JENKINS-67237_2nd_attempt

basil committed:
Clean up formatting

basil reviewed (approved):
I pushed a minor commit to clean up the formatting. Thanks for taking care of this!

---

This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

basil committed:
Merge branch 'master' into JENKINS-67237_2nd_attempt

basil commented:
The `skip-changelog` label was added by mistake.
